compiler/ilgen/OMRIlBuilder.cpp

```diff
@@ -2745,3 +2757,34 @@ OMR::IlBuilder::WhileDoLoop(const char *whileCondition, TR::IlBuilder **body, TR
 
    AppendBuilder(done);
    }
+
+void *
+OMR::IlBuilder::client()
+   {
+   if (_client == NULL && _clientAllocator != NULL)
+      _client = _clientAllocator(static_cast<TR::IlBuilder *>(this));
+   return _client;
+   }
+
+void *
+OMR::IlBuilder::JBCase::client()
+   {
+   if (_client == NULL && _clientAllocator != NULL)
```

Comment:
Clients are supposed to allocate these objects by calling `MakeCase()` (and `MakeCondition()` for JBCondition). On the client side, `OMR::JitBuilder::IlBuilder::MakeCase()` will call `TR::IlBuilder::MakeCase()`. Then, `client()` is called on the returned instance of `TR::IlBuilder::JBCase` to get the corresponding client-side object, which is finally returned to the client.

Does that make sense?

Reply:
it does, but then what use in a _clientAllocator ? I expected that only to be useful if 1) the implementation needed to allocate a client object and 2) we wanted clients to be able to provide their own "subclass" of the client object type ?

Reply:
in the scenario you describe, the client has already provided a client object so there shouldn't be any need to allocate?

Reply:
Not quite. Because `OMR::JitBuilder::IlBuilder::MakeCase()` calls `TR::IlBuilder::MakeCase()`, the implementation-side objects will get allocated first. The client-side objects are allocated when `client()` is called, which implies the existence of an implementation object.